🦜 🏴‍☠️ Implement NodePiece representation and model

mberr opened the pull request:
This is a first draft to add NodePiece representations to pykeen.

For now, it uses a simple variant of it, where each entity is represented by `k` randomly chosen incident relations.

* Official Code: https://github.com/migalkin/NodePiece
* Paper: https://arxiv.org/abs/2106.12144

![One_Piece,_Volume_61_Cover_(Japanese)-1](https://user-images.githubusercontent.com/5069736/140743316-3d03e4b2-a44d-4fda-9ebc-92cfb73ab449.jpg)

mberr committed:
add node pieces representation

mberr committed:
add test case for node piece representations

mberr committed:
fix reduction

mberr committed:
remove unused import

trigger ci

mberr committed:
fix name & add placeholder for citation

trigger ci

mberr committed:
re-use embedding specification

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,103 @@ def forward(
+            fill_value=2 * triples_factory.num_relations,  # padding ID
+        )
+        for e, rs in e2r.items():
+            assignment[e] = torch.as_tensor(data=random.choices(population=list(rs), k=k), dtype=torch.long)
```
can torch not handle random sampling? this introduces yet another random thing we'd have to control for reproducibility :/

mberr replied:
```suggestion
            rs = torch.as_tensor(data=list(rs), dtype=torch.long)
            assignment[e] = rs[torch.randint(high=rs.numel(), size=(k,))]
```

cthoyt replied:
Done in f771e7ae

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,103 @@ def forward(
+            raise NotImplementedError("Triples factories with inverse triples are currently not supported.")
+
+        # create token representations
+        tokens = token_representation.make(
```
Can we go deeper? can token representations be any arbitrary RepresentationModule? or does it have to be embeddings?

mberr replied:
It can be any. I just did not find a nice representation resolver to delegate this to :wink:

cthoyt replied:
I think that would be *very* difficult

cthoyt replied:
We should allow it to pass an instantiated representation in case we want to build a model whose `__init__` takes care of instantiating something more exotic, like R-GCN representations!!

mberr replied:
Although, notice that the token representations are (at least in this variant) only used for *relation* representations.

mberr replied:
done in 70bf759

mberr replied:
> I think that would be very difficult

why actually?

cthoyt replied:
because the arguments are potentially very different and need to be handled differently

I could be wrong though

mberr replied:
In the easiest case, we would not handle the parameters at all, but require the user to provide them properly.

```python
def __init__(
  self,
  representation: HintOrType[Representations],
  representation_kwargs: Optional[Mapping[str, Any]] = None,
  ...,
):
  representation = representation_resolver.make(representation, representation_kwargs)
```

cthoyt replied:
I think this is moot at the moment since the way these are constructed for the relation is now pretty specific.

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,103 @@ def forward(
+        :param token_representation:
+            the token representation specification
+        :param aggregation:
+            aggregation of multiple token representations to a single entity representation
```
besides `torch.mean`, what other things might go here? Is `torch.sum` reasonable? Or anything else? Please add to docs :)

mberr replied:
The simple ones would be sum/mean/max.

We can also have trainable aggregations, e.g., `MLP(mean(MLP(tokens)))` (cf. [DeepSets](https://arxiv.org/abs/1703.06114)).

In theory, we could also have aggregations which result in differently shapes output, e.g. a concatenation of all token embeddings resulting in shape `(k * d,)`, but for this, we would need to adjust the code.

I tried to keep it simple for now :slightly_smiling_face:

cthoyt replied:
I'll add this to the docs

cthoyt replied:
Done in c614a485

mberr replied:
For reference: the paper uses either
 1) MLP that takes as input a concatenated hash vector R1×(k+m)d projecting it down to Rd;  or
2) Transformer encoder [42] with average pooling that takes as input an original sequence R(k+m)×d

mberr committed:
add node piece model

mberr committed:
Update src/pykeen/nn/emb.py

mberr committed:
allow custom output shape

cthoyt committed:
Update emb.py

mberr committed:
allow provision of pre-instantiated token repr.

mberr committed:
Merge branch 'node-pieces' of github.com:pykeen/pykeen into node-pieces

mberr committed:
export model and add tests

trigger ci

mberr commented on tests/test_nn/test_emb.py:
```diff
@@ -113,6 +113,30 @@ def _pre_instantiation_hook(self, kwargs: MutableMapping[str, Any]) -> MutableMa
+
+    def _pre_instantiation_hook(self, kwargs: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
+        kwargs = super()._pre_instantiation_hook(kwargs)
+        kwargs["triples_factory"] = generate_triples_factory(
+            num_entities=self.num_entities,
+            num_relations=self.num_relations,
+            num_triples=self.num_triples,
+            create_inverse_triples=False,
+        )
```
This is duplicated code to RGCNRepresentation tests

cthoyt replied:
It's sort of duplicated, but abstracting this would actually create more code at this point. I'll leave a note for later, just in case.

cthoyt replied:
TODO added (for a later PR) in 7e1813ef

cthoyt committed:
Update citation and readme

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,125 @@ def forward(
+            aggregation of multiple token representations to a single entity representation. By default,
+            this uses :func:`torch.mean`. It could also use other aggregations like :func:`torch.sum`,
+            :func:`torch.max`, or even trainable aggregations e.g., ``MLP(mean(MLP(tokens)))``
+            (cf. DeepSets).
```
reference for DeepSets?

mberr replied:
e.g. https://arxiv.org/abs/1703.06114

mberr replied:
maybe this one is better: https://papers.nips.cc/paper/2017/hash/f22e4747da1aa27e363d86d40ff442fe-Abstract.html

cthoyt replied:
Done in 5bbc4bb3

cthoyt committed:
Remove missing import

cthoyt committed:
Update docs

cthoyt committed:
Remove unnecessary kwargs

cthoyt committed:
Fix last mypy

Trigger CI

cthoyt committed:
Add citation

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,127 @@ def forward(
+class NodePieceRepresentation(RepresentationModule):
+    """
+    Basic implementation of node piece decomposition [galkin2021]_.
+
```
request for fancy math documentation of what this is doing

mberr replied:
7e767bc 

this does not necessarily match the paper's description, but I found it more intuitive given that you may not have read the paper until then.

cthoyt replied:
Excellent. I strive for the PyKEEN documentation to be educational, not just a way to make stuff look good. @migalkin already did an excellent job of that in his paper 💯

cthoyt committed:
Update typing in tests

cthoyt committed:
Update test_emb.py

mberr committed:
add math for node piece representation

mberr committed:
Merge branch 'node-pieces' of github.com:pykeen/pykeen into node-pieces

mberr committed:
add concat + MLP version from paper

mberr committed:
add comment

trigger ci

mberr committed:
add comment

trigger ci

mberr committed:
fix mypy

trigger ci

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1072,6 +1072,27 @@ def forward(
+        )
+
+    def forward(xs: torch.FloatTensor, dim: int) -> torch.FloatTensor:  # noqa: D102
+        assert dim == -2
```
can you add a more substantial error message that says why it has to be -2

mberr replied:
I moved it to the model file; this is only a private class, which is not thought to be visible from the outside, and only serves as a convenience method to get the configuration from the paper.

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,177 @@ def forward(
+        super().__init__(max_id=triples_factory.num_entities, shape=shape or token_representation.shape)
+
+        # normalize aggregation
+        if aggregation is None:
```
```suggestion
        if aggregation is None or aggregation == "mean":
```

mberr replied:
I moved everything string related to the model class, and kept the Representation clean from these shortcuts.

mberr committed:
move MLP to model class, keep repr. generic

cthoyt commented on src/pykeen/models/unimodal/node_piece.py:
```diff
@@ -60,10 +104,21 @@ def __init__(
+                num_tokens=num_tokens,
+                embedding_dim=embedding_specification.embedding_dim,
+            )
+            aggregation = self.mlp.forward
```
I think just doing `self.mlp` is better since this is a callable and this is the preferred way to use the forward()

cthoyt replied:
done in afc873b4

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,133 @@ def forward(
+        x = self.tokens(token_ids)
+
+        # aggregate
+        x = self.aggregation(x, -2)
```
Add docs on what the `-2` means (typically, the dimension)

mberr replied:
I think we should just get rid of it (by applying functools.partial to torch.mean)

cthoyt replied:
but how should someone bring their own function then?

cthoyt committed:
Update node_piece.py

cthoyt committed:
Make ratio parametrizable

cthoyt committed:
Add my encoding header

cthoyt committed:
Reorganize docs and make class private

cthoyt committed:
Add tests for nodepiece with MLP aggregation

migalkin committed:
Fix for embedding_dim in the initialization of the aggregation encoder

migalkin committed:
Removed self.mlp assignment which happens before super().__init__() and causes the AttributeError

cthoyt commented on src/pykeen/models/unimodal/node_piece.py:
```diff
@@ -118,7 +118,7 @@ def __init__(
         if aggregation == "mlp":
             # needs to be assigned to attribute to make sure that the trainable parameters are part of the model
             # parameters
-            node_piece_kwargs["aggregation"] = self.mlp = _ConcatMLP(
```
looks like you missed the docstring above saying why we had to do it like this

cthoyt commented on src/pykeen/models/unimodal/node_piece.py:
```diff
@@ -0,0 +1,136 @@
+        if aggregation == "mlp":
+            # needs to be assigned to attribute to make sure that the trainable parameters are part of the model
+            # parameters
+            node_piece_kwargs["aggregation"] = _ConcatMLP(
```
@migalkin removed the assignment to self.mlp, so maybe he knows something we don't?

mberr replied:
716f343 mentions an instantiation error, since this was done before super -> should be saved in a local variable and assigned to an attribute after the super call

cthoyt committed:
Update node_piece.py

cthoyt committed:
Don't worry about setting self.mlp

since it gets set into something else that alseo gets set it gets recursively added

cthoyt committed:
Update node_piece.py

cthoyt committed:
Fix typing and pass flake8/mypy

Trigger CI

mberr committed:
Merge branch 'master' into node-pieces

mberr committed:
Merge branch 'master' into node-pieces

mberr committed:
add support for triples factories with inv triples

mberr committed:
share relation parameters

trigger ci

mberr committed:
fix filtering

trigger ci

mberr committed:
emit warning for missing inverse relations

trigger ci

cthoyt committed:
Merge branch 'master' into node-pieces

cthoyt committed:
Update emb.py

Trigger CI

cthoyt committed:
Merge branch 'master' into node-pieces

cthoyt committed:
Update README.md

cthoyt committed:
Merge branch 'master' into node-pieces

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Make kwargs more explicit and add several questions

cthoyt committed:
Revert aggreation setting

cthoyt committed:
update docs

migalkin commented on src/pykeen/models/unimodal/node_piece.py:
```diff
@@ -0,0 +1,158 @@
+            aggregation of multiple token representations to a single entity representation. By default,
+            this uses :func:`torch.mean`. It could also use other aggregations like :func:`torch.sum`,
+            :func:`torch.max`, or even trainable aggregations e.g., ``MLP(mean(MLP(tokens)))``
+            (cf. DeepSets from [zaheer2017]_) if given value ``"mlp"``.
```
The current `_ConcatMLP` is not DeepSets :)

The idea of DeepSets is to project each set member independently through some encoder, then aggregate (like with mean) and then pass through another FF net. It would look like this:
```
enc1 = nn.Sequential(
    nn.Linear(embedding_dim, embedding_dim),
    nn.ReLU(),
    nn.Linear(embedding_dim, embedding_dim)
)

enc2 = nn.Sequential(nn.Linear(embedding_dim, embedding_dim), nn.ReLU(), nn.Linear(embedding_dim, output_dim))
```
and in forward pass:
```
# x: shape (bs, num_elements, embedding_dim)
x = enc1(x)               # the same shape (bs, num_elements, embedding_dim)
x = torch.mean(-2)  # here we do the aggregation to (bs, embedding_dim)
x = enc2(x)               # final projection keeping (bs, output_dim)
```

It can be added as an option along with mlp though

mberr replied:
The correct docstring somehow got lost during the refactoring :sweat_smile: 

here it was still correct: https://github.com/pykeen/pykeen/pull/621#discussion_r744711500

mberr committed:
sample without replacement

mberr committed:
fix refactoring artifact

cthoyt committed:
Address aggregation CLI parameter

cthoyt committed:
Merge branch 'master' into node-pieces

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,136 @@ def forward(
+        """
+        mapped_triples = triples_factory.mapped_triples
+        if triples_factory.create_inverse_triples:
+            # inverse triples are created afterwards implicitly
```
I would rather have the code explicitly disallow triples factories with create_inverse_triples rather than writing code to try and undo the operation

mberr replied:
This would prohibit using inverse triples for computing triple scores. The reference implementation always uses inverse trples, and also the implementation of the NodePiece uses `2 * R + 1` relation embeddings (with an additional padding relation embdding which is not part of any other model so far).

mberr commented on src/pykeen/models/unimodal/node_piece.py:
```diff
@@ -137,6 +137,12 @@ def __init__(
+        elif aggregation == "sum":
+            aggregation = torch.sum
+        elif aggregation == "max":
+            aggregation = torch.max
```
Alternative:
```suggestion
        else:
            aggregation = getattr(torch, aggregation)
```

cthoyt replied:
I like being a bit more explicit. This alternative allows the code to do some wrong things, like if you put "tan"

mberr replied:
It would also allow to e.g. use [logsumexp](https://pytorch.org/docs/stable/generated/torch.logsumexp.html) (= soft version of maximum) without requiring any code changes.

cthoyt replied:
I still think it's better to explicitly handle these, so we can better document what users should put in

mberr committed:
bring back shape parameter

mberr committed:
fix name collision

mberr committed:
collate code

mberr committed:
extract helper function

cthoyt committed:
Update aggregations list

mberr committed:
move aggregation resolution downwards

cthoyt committed:
Merge branch 'master' into node-pieces

mberr committed:
Merge branch 'node-pieces' of github.com:pykeen/pykeen into node-pieces

mberr committed:
Merge branch 'node-pieces' of github.com:pykeen/pykeen into node-pieces

cthoyt committed:
Remove invalid kwarg

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -1068,3 +1070,157 @@ def forward(
+
+            The aggregation takes two arguments: the (batched) tensor of token representations, in shape
+            ``(*, num_tokens, *dt)``, and the index along which to aggregate.
+        :param shape:
```
Do we have a compelling use case for this? Otherwise I think it's quite a confusing setting

mberr replied:
citing from a few lines above 😉 
https://github.com/pykeen/pykeen/blob/4ee90a30ddad81051122923a6c5ee3599b84e894/src/pykeen/models/unimodal/node_piece.py#L113-L114

cthoyt replied:
lol RTFD, charlie 😆

mberr committed:
make aggregation resolution more explicit

mberr committed:
add docstring

trigger ci

mberr committed:
Merge branch 'node-pieces' of github.com:pykeen/pykeen into node-pieces

mberr committed:
trigger ci

mberr committed:
improve documentation & remove aggregation const.s

trigger ci

mberr committed:
revert change to constants

trigger ci

mberr committed:
update doc

trigger ci

mberr committed:
slightly generalize _ConcatMLP

support floating point ratios

trigger ci

mberr commented on src/pykeen/models/unimodal/node_piece.py:
```diff
@@ -0,0 +1,159 @@
+                "The provided triples factory does not create inverse triples. However, for the node piece"
+                "representations inverse relation representations are required. Thus, the implicitly created inverse "
+                "relations are only trained via the NodePiece mechanism, but not as part of the 'normal' training.",
+            )
```
@cthoyt I would also support to raise an exception here (since the models here are just a convenient way to compose the more general `ERModel` according to some papers' configurations).

cthoyt replied:
Yes I think raising an exception rather than trying to compensate in a complicated way is a good idea

mberr replied:
ac97a56

mberr committed:
fix format

trigger ci

mberr commented on src/pykeen/models/unimodal/node_piece.py:
```diff
@@ -0,0 +1,159 @@
+        num_tokens: int,
+        embedding_dim: int,
+        dropout: float = 0.1,
+        ratio: Union[int, float] = 2,
```
@cthoyt this parameter is not exposed via the `NodePiece`model, and the class is not intended to be used outside. So maybe we can drop the parameter again?

cthoyt replied:
That's true, but I like the idea of it being changable/documented through the kwargs

mberr committed:
add num_tokens to docstring

trigger ci

mberr committed:
re-apply black

trigger ci

mberr committed:
add test for score_r, and fix typos

mberr committed:
re-use code from test case

mberr committed:
fix typo in nbase

mberr committed:
trigger ci

mberr committed:
raise error instead of warn

mberr committed:
implement inverse triples creation for CLI test

also fix create_inverse_triples for NodePiece test

trigger ci

mberr committed:
hide padding relation to enable relation scoring

trigger ci

mberr commented on tests/cases.py:
```diff
@@ -931,7 +931,7 @@ def test_score_t(self) -> None:
             scores = self.instance.score_t(batch)
         except NotImplementedError:
-            self.fail(msg="Score_o not yet implemented")
+            self.fail(msg="score_t not yet implemented")
```
this typo is not really part of the PR

mberr commented on tests/cases.py:
```diff
@@ -950,7 +968,7 @@ def test_score_h(self) -> None:
             scores = self.instance.score_h(batch)
         except NotImplementedError:
-            self.fail(msg="Score_s not yet implemented")
+            self.fail(msg="score_h not yet implemented")
```
same here

mberr committed:
fix darglint

trigger ci

mberr committed:
fix negative index not being supported by module

trigger ci

migalkin reviewed (approved):
The implementation works 🎉 
Exposing the `ratio` param from the MLP encoder sounds like a good idea (with the default value 2), otherwise everything looks ready!

mberr committed:
add warning

mberr committed:
update docstring

trigger ci

mberr committed:
skip score_r shape verification if inv. relations

it is not yet clear what would be the desired output shape

mberr committed:
add TODO

trigger ci

mberr committed:
add test for SubsetRepresentationModule

trigger ci

cthoyt reviewed (approved):
:shipit:
